asset-registry/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_asset())]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::do_register_asset(metadata, asset_id)
+		}
+
+		#[allow(clippy::too_many_arguments)]
+		#[pallet::weight(T::WeightInfo::update_asset())]
+		#[transactional]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			decimals: Option<u32>,
+			name: Option<Vec<u8>>,
+			symbol: Option<Vec<u8>>,
+			existential_deposit: Option<T::Balance>,
+			location: Option<Option<VersionedMultiLocation>>,
+			additional: Option<T::CustomMetadata>,
+		) -> DispatchResult {
+			T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::do_update_asset(
+				asset_id,
+				decimals,
+				name,
+				symbol,
+				existential_deposit,
+				location,
+				additional,
+			)?;
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Register a new asset
+	pub fn do_register_asset(
+		metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		asset_id: Option<T::AssetId>,
+	) -> DispatchResult {
+		let (asset_id, metadata) = T::AssetProcessor::pre_register(asset_id, metadata)?;
+
+		Self::do_register_asset_without_asset_processor(metadata.clone(), asset_id.clone())?;
+
+		T::AssetProcessor::post_register(asset_id, metadata)?;
+
+		Ok(())
+	}
+
+	/// Like do_register_asset, but without calling pre_register and
+	/// post_register hooks.
+	/// This function is useful in tests but it might also come in useful to
+	/// users.
+	pub fn do_register_asset_without_asset_processor(
+		metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		asset_id: T::AssetId,
+	) -> DispatchResult {
+		Metadata::<T>::try_mutate(&asset_id, |maybe_metadata| -> DispatchResult {
+			// make sure this asset id has not been registered yet
+			ensure!(maybe_metadata.is_none(), Error::<T>::ConflictingAssetId);
+
+			*maybe_metadata = Some(metadata.clone());
+
+			if let Some(ref location) = metadata.location {
+				Self::do_insert_location(asset_id.clone(), location.clone())?;
+			}
+
+			Ok(())
+		})?;
+
+		Self::deposit_event(Event::<T>::RegisteredAsset { asset_id, metadata });
+
+		Ok(())
+	}
+
+	pub fn do_update_asset(
+		asset_id: T::AssetId,
+		decimals: Option<u32>,
+		name: Option<Vec<u8>>,
+		symbol: Option<Vec<u8>>,
+		existential_deposit: Option<T::Balance>,
+		location: Option<Option<VersionedMultiLocation>>,
+		additional: Option<T::CustomMetadata>,
+	) -> DispatchResult {
+		Metadata::<T>::try_mutate(&asset_id, |maybe_metadata| -> DispatchResult {
+			match maybe_metadata {
+				None => Err(Error::<T>::AssetNotFound.into()),
```

Comment:
replace with below, and remove match ase. also replace `(*metadata)` with `metadata`
```rust
let metadata = maybe_metadata.as_mut().ok_or(Error::<T>::AssetNotFound)?;
```